pkg/socketlb/cgroup.go

```diff
@@ -0,0 +1,240 @@
+		return nil, fmt.Errorf("detach program: %w", err)
+	}
+
+	l, err := link.AttachCgroup(link.CgroupOptions{
+		Path:    cgroupRoot,
+		Attach:  attachTypes[name],
+		Program: prog,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("attach program %s: %w", name, err)
+	}
+
+	return l, nil
+}
+
+// detachCgroup detaches a program with the given name from cgroupRoot. Attempts
+// to open a pinned link with the given name from directory pinPath first,
+// falling back to PROG_DETACH if no pin is present.
+func detachCgroup(name, cgroupRoot, pinPath string) error {
+	pin := filepath.Join(pinPath, name)
+	l, err := link.LoadPinnedLink(pin, &ebpf.LoadPinOptions{})
+	if err == nil {
+		if err := l.Unpin(); err != nil {
+			return fmt.Errorf("unpin link %s: %w", pin, err)
+		}
+		if err := l.Close(); err != nil {
+			return fmt.Errorf("close link %s: %w", name, err)
+		}
+		return nil
+	}
+
+	// No bpf_link pin found, search and detach prog_attach progs.
+	log.Debugf("no pinned link '%s', querying cgroup", pin)
+	err = searchAndDetach(name, cgroupRoot, nil)
+	// Treat detaching unsupported attach types as successful.
+	if errors.Is(err, link.ErrNotSupported) {
+		return nil
+	}
+	return err
+}
+
+// tryPin pins the link to maintain it after agent exit. l is closed after
+// a successful pin operation.
+//
+// Returns ErrNotSupported if l was attached using PROG_ATTACH.
+func tryPin(l link.Link, name, pinPath string) error {
+	err := l.Pin(filepath.Join(pinPath, name))
+	if errors.Is(err, link.ErrNotSupported) {
+		log.Debugf("program %s was attached using PROG_ATTACH", name)
+	}
+	if err != nil {
+		return fmt.Errorf("pin link: %w", err)
+	}
+
+	log.Debugf("program %s attached using bpf_link", name)
+	return l.Close()
+}
+
+// searchAndDetach detaches all programs attached to cgroupRoot with the
+// corresponding attach type for a program named 'name'. If skip is given,
+// that program will be skipped during the operation.
+func searchAndDetach(name, cgroupRoot string, skip *ebpf.Program) error {
+	fd, err := os.Open(cgroupRoot)
+	if err != nil {
+		return fmt.Errorf("open cgroup %s: %w", cgroupRoot, err)
+	}
+	defer fd.Close()
+
+	attach := attachTypes[name]
+	ids, err := link.QueryPrograms(link.QueryOptions{
+		Path:   cgroupRoot,
+		Attach: attach,
+	})
+	// We know the cgroup root exists, so EINVAL will likely mean querying
+	// the given attach type is not supported.
+	if errors.Is(err, unix.EINVAL) {
+		err = fmt.Errorf("%s: %w", err, link.ErrNotSupported)
+	}
+	if err != nil {
+		return fmt.Errorf("query cgroup %s for type %s: %w", cgroupRoot, attach, err)
+	}
+
+	if len(ids) == 0 {
+		log.Debugf("no programs in cgroup %s with attach type %s and name %s", cgroupRoot, attach, name)
+		return nil
+	}
+
+	var sid ebpf.ProgramID
+	if skip != nil {
+		info, err := skip.Info()
+		if err != nil {
+			return fmt.Errorf("get program info: %w", err)
+		}
+
+		id, ok := info.ID()
+		if !ok {
+			return fmt.Errorf("kernel does not support getting program id")
```

Comment:
Can we do something more in this case?

I feel like the execution path to get to here goes very deep into details, completing a lot of the loading operations, only to get to here and fail out, leaving the system in an inconsistent state. Can we catch these earlier, for instance by querying for support for such operations during startup and then tailoring the cgroup prog attach logic based on whether we can actually detach programs or not?

Or am I looking to deeply into this and we never expect this to be hit?

Reply:
My thought here is that in this error case, Cilium manages to load the new cgroup programs (possibly via `exclusiveCgroupAttach()` which already went through this function the first time), then attempt a pin, but pinning fails so then the execution flow comes down to here, and then rather than attempting to delete all programs and attach just the one we want, this function gives up here, *maybe* leaving the system with multiple programs attached to the cgroup?

Reply:
(I checked how the callers handle this sort of error condition, and it looks like it will just get propagated all the way up, so the eventual state in terms of attached cgroup programs seems undeterministic)